client/requester.go

```diff
@@ -0,0 +1,292 @@
+		if err != nil {
+			return nil, err
+		}
+		*ws = conn
+		return nil, nil
+	}
+
+	// Is the result expecting a caller managed body reader?
+	if bodyReader, ok := result.(*BodyReader); ok {
+		httpResp, err := br.raw(ctx, opts.Method, opts.Path, opts.Query, opts.Headers, opts.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		*bodyReader = httpResp.Body
+		return nil, nil
+	}
+
+	// This is a normal sync or async server request
+	httpResp, err := br.rawWithRetry(ctx, opts.Method, opts.Path, opts.Query, opts.Headers, opts.Body)
+	if err != nil {
+		return nil, err
+	}
+	defer httpResp.Body.Close()
+
+	// Get the client decoder to extract what it needs before we proceed
+	reqResp, err := br.decoder(ctx, httpResp, opts, result)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sanity check sync and async requests
+	if opts.Async == false {
+		if reqResp.Type != "sync" {
+			return nil, fmt.Errorf("expected sync response, got %q", reqResp.Type)
+		}
+	} else {
+		if reqResp.Type != "async" {
+			return nil, fmt.Errorf("expected async response for %q on %q, got %q", opts.Method, opts.Path, reqResp.Type)
+		}
+		if reqResp.StatusCode != 202 {
+			return nil, fmt.Errorf("operation not accepted")
+		}
+		if reqResp.Change == "" {
+			return nil, fmt.Errorf("async response without change reference")
+		}
+	}
+
+	return reqResp, nil
+}
+
+// SocketNotFoundError is the error type returned when the client fails
+// to find a unix socket at the specified path.
+type SocketNotFoundError struct {
+	// Err is the wrapped error.
+	Err error
+
+	// Path is the path of the non-existent socket.
+	Path string
+}
+
+func (s SocketNotFoundError) Error() string {
+	if s.Path == "" && s.Err != nil {
+		return s.Err.Error()
+	}
+	return fmt.Sprintf("socket %q not found", s.Path)
+}
+
+func (s SocketNotFoundError) Unwrap() error {
+	return s.Err
+}
+
+// decodeWithNumber decodes input data using json.Decoder, ensuring numbers are preserved
+// via json.Number data type. It errors out on invalid json or any excess input.
+func decodeWithNumber(r io.Reader, value interface{}) error {
+	dec := json.NewDecoder(r)
+	dec.UseNumber()
+	if err := dec.Decode(&value); err != nil {
+		return err
+	}
+	if dec.More() {
+		return fmt.Errorf("cannot parse json value")
+	}
+	return nil
+}
+
+func unixDialer(socketPath string) func(string, string) (net.Conn, error) {
+	return func(_, _ string) (net.Conn, error) {
+		_, err := os.Stat(socketPath)
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, &SocketNotFoundError{Err: err, Path: socketPath}
+		}
+		if err != nil {
+			return nil, fmt.Errorf("cannot stat %q: %w", socketPath, err)
+		}
+
+		return net.Dial("unix", socketPath)
+	}
+}
+
+type doer interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// RequestError is returned when there's an error processing the request.
+type RequestError struct{ error }
+
+func (e RequestError) Error() string {
+	return fmt.Sprintf("cannot build request: %v", e.error)
+}
```

Comment:
```suggestion
type RequestError struct{ error }

func (e RequestError) Error() string {
	return fmt.Sprintf("cannot build request: %v", e.error)
}

func (e RequestError) Unwrap() error {
	return e.error
}
```

Reply:
Existing code. Happy to change if I need to.